theories/crypto/KeyEncapsulationMechanismsROM.eca

```diff
@@ -0,0 +1,1476 @@
+    var r : bool;
+    
+    RO.init();
+    
+    r <@ Correctness(S(RO)).main();
+    
+    return r;
+  }
+}.
+
+
+(* Security in ROM *)
+(* Attacker capabilities/models in ROM *)
+(* 
+  Chosen-Plaintext Attacks (CPA) in ROM.
+  The adversary is given the considered public key and, hence, 
+  is able to produce encapsulations, i.e., (symmetric) key/ciphertext pairs. 
+  (Typically, this means the adversary can construct ciphertexts
+   corresponding to chosen (symmetric) keys.)
+*)
+
+(* 
+  non-adaptive Chosen-Ciphertext Attacks (CCA1) in ROM.
+  The adversary is given the considered public key and access to a decryption oracle
+  *before* the stage in which it is expected to distinguish/return a break.
+  Hence, the adversary is able to produce encapsulations
+  *and* query for decryptions of chosen ciphertexts.
+*)
+(** Interface for oracles employed in CCA1 security games in ROM **)
+module type Oracles_CCA1i_ROM (RO : RandomOracle) (S : Scheme) = {
+  proc init(sk_init : sk_t) : unit
+  proc decaps(c : ctxt_t) : key_t option
+}.
+
+
+(* 
+  adaptive Chosen-Ciphertext Attacks (Traditional: CCA2, Modern : CCA) in ROM.
+  The adversary is given the considered public key and access to a decryption oracle throughout.  
+  Hence, the adversary is able to produce encapsulations
+  *and* query for decryptions of chosen ciphertexts (potentially barring ciphertexts
+  that are part of the challenge).
+  Traditionally, this was analogous to CCA2 security for PKE schemes, meaning there were
+  two adversary stages: one before receiving the challenge (given a public key and access to a 
+  non-restricted decapsulation oracle), and one after receiving the challenge (given access to a
+  restricted decapsulation oracle, i.e., one that prohibited querying the challenge).
+  Over time, the formalization shifted toward only considering the second adversary stage 
+  (provding the public key(s) to this stage as well).
+  Here, we denote the traditional one by CCA2 (as we do for PKE schemes), and the modern one by CCA.
+*)
+(** Interface for oracles employed in CCA2 (CCA) security games in ROM **)
+module type Oracles_CCA2i_ROM (RO : RandomOracle) (S : Scheme) = {
+  proc init(sk_init : sk_t, c'_init : ctxt_t) : unit
+  proc decaps(c : ctxt_t) : key_t option
+}.
+
+
+(* 
+  One-Wayness (OW) in ROM.
+  The adversary is asked to produce the (symmetric) key
+  encapsulated by a given ciphertext.
+*)
+(*
+  One-Wayness under Chosen-Plaintext Attacks (OW-CPA) in ROM.
+  In a CPA setting, the adversary is asked to produce the (symmetric) key
+  encapsulated by a given ciphertext.
+*)
+(** Adversary class considered for OW-CPA in ROM **)
+module type Adv_OWCPA_ROM (RO : RandomOracle) = {
+  include Adv_OWCPA
+}.
+
+(** OW-CPA security game in ROM **)
+module OW_CPA_ROM (RO : RandomOraclei) (S : Scheme_ROM, A : Adv_OWCPA_ROM) = {
+  proc main() : bool = {
+    var r : bool;
+    
+    RO.init();
+    
+    r <@ OW_CPA(S(RO), A(RO)).main();
```

Comment:
One could also provide a `type d_in_t.` parameter to the theory, to allow instantiation with distinguishers that take in public parameters. (For example, in hybrid arguments, or when a trusted setup outputs public parameters.)

Reply:
Yes, so this is also something I was unsure of, and also hinges on the answer to "do we want to force a dependency on the PROM theory" question. If we don't want to force a dependency on PROM, I don't think it makes sense to also have PROM-specific stuff that doesn't make sense nor is used in the KEM-ROM theory.  If we do want to force a dependency, I suppose it depends: we could explicitly instantiate `d_in_t` (and also have it as an abstract type in the Top-level KEM-ROM theory), or we could only instantiate `in_t` and `out_t` (since those are all that is used/referred to in the KEM-ROM theory itself). The latter, which is currently the case, expects people using the library and actually want to use things from PROM to know what they are doing and also instantiate d_in_t and the like  when instantiating KEM-ROM (by going a layer deeper in the cloning, e.g., `type F_RO.d_in_t <- ...`).

Reply:
I think this fits with the broader discussions of "deep ugly overly-parameterised generic theories with cleanly defined instantiations".